mgmt: hawkbit: make server address and port settable during runtime

maass-hamburg opened the pull request:
In situations, where the address and the port of the hawkbit server is not known during build, it should be possible to set it during runtime.

maass-hamburg commented on subsys/mgmt/hawkbit/hawkbit.c:
```diff
@@ -598,6 +632,25 @@ int hawkbit_init(void)
 	LOG_DBG("Action id: current %d", action_id);
 
+	if (!strcmp(hawkbit_server, "")) {
+		if (strcmp(CONFIG_HAWKBIT_SERVER, "")) {
```
this should only be executed when initializing hawkbit and not everytime hawkbit_set_server_addr is executed. it is there to use the value from Kconfig, if the address havn't been set by `hawkbit_set_server_addr`

ycsin replied:
It'd be nice if this internal variable is only modified from a single function

maass-hamburg commented on subsys/mgmt/hawkbit/hawkbit.c:
```diff
@@ -598,6 +632,25 @@ int hawkbit_init(void)
+		}
+	}
+
+	if (!strcmp(hawkbit_server_port, "")) {
+		if (strcmp(CONFIG_HAWKBIT_PORT, "")) {
+			strncpy(hawkbit_server_port, CONFIG_HAWKBIT_PORT,
+				sizeof(hawkbit_server_port));
+		} else {
+			LOG_ERR("no valid hawkbit server port found");
+			return -EFAULT;
+		}
+	}
```
Same as before and  `hawkbit_set_server_port` expects a uint16_t and not a string, therefore it can't be used with `CONFIG_HAWKBIT_PORT`

maass-hamburg committed:
mgmt: hawkbit: change to K_SEM_DEFINE

Define and initialize probe_sem with K_SEM_DEFINE

Signed-off-by: Fin Maaß <f.maass@vogl-electronic.com>

ycsin commented on subsys/mgmt/hawkbit/hawkbit.c:
```diff
@@ -65,6 +66,8 @@ static uint32_t poll_sleep = (CONFIG_HAWKBIT_POLL_INTERVAL * 60 * MSEC_PER_SEC);
 #endif
 
+static char hawkbit_server[DNS_MAX_NAME_SIZE + 1];
+static char hawkbit_server_port[6];
```
I'm not sure if there's a definition for the max value of a port:
```suggestion
static char hawkbit_server_port[sizeof(STRINGIFY(__UINT16_MAX__))];
```

ycsin reviewed (commented):
I can see how this can be useful for use cases that need to configure the server/port to something else, some thoughts:
 - I think it probably make sense to have a struct (i.e. `hawkbit_config`) that stores everything it needs to connect to a server (addr, port & security token), and can be configured with a single function.
 - These configs can be stored in the nvs filesystem, similar to the action ID - if there isn't an entry in the nvs, fallback to the Kconfig values
   - You can probably also convert the existing implementation with `nvs` to use the `setting` subsys while you are here if you like
 - Would be nice to also update the [hawkbit shell](https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/mgmt/hawkbit/shell.c) to make use of this new functionality
 - Currently the additional functions and buffer are there even if the users are not going to configure them in runtime, would be nice if this is configurable via Kconfig

maass-hamburg committed:
mgmt: hawkbit: change Kconfig port to int

Ports set by Kconfig are mostly defined
as a int. This changes it for hawkbit too.

Signed-off-by: Fin Maaß <f.maass@vogl-electronic.com>

maass-hamburg committed:
mgmt: hawkbit: set server address, port and auth_token during runtime

In situations, where the address, the port and the auth token of the
hawkbit server are not known during build, it should be possible to set
it during runtime.

Signed-off-by: Fin Maaß <f.maass@vogl-electronic.com>

maass-hamburg committed:
mgmt: hawkbit: extend shell to settings

This commit extents the functions
of the hawkbit shell to set the settings

Signed-off-by: Fin Maaß <f.maass@vogl-electronic.com>

ycsin commented on subsys/mgmt/hawkbit/hawkbit.c:
```diff
@@ -200,6 +196,40 @@ static const struct json_obj_descr json_dep_fbk_descr[] = {
+	return -ENOENT;
+}
+
+int hawkbit_settings_export(int (*cb)(const char *name, const void *value, size_t val_len))
```
can this be `static`?

ycsin commented on subsys/mgmt/hawkbit/hawkbit.c:
```diff
@@ -216,6 +225,55 @@ static int hawkbit_settings_set(const char *name, size_t len, settings_read_cb r
 		return rc;
 	}
 
+	size_t next_len = settings_name_next(name, &next);
```
I think it probably makes sense to use just one level? (i.e. `hawkbit/server_addr`, `hawkbit/server_port`, `hawkbit/action_id`) then you will be able to literally copy paste the same thing over and over again, be it `settings_name_steq` or `strncmp(..)`

seems like we can factorize this function quite abit

ycsin commented on subsys/mgmt/hawkbit/hawkbit.c:
```diff
@@ -598,6 +673,73 @@ static void hawkbit_dump_deployment(struct hawkbit_dep_res *d)
+		LOG_DBG("hawkbit server port already correctly configured");
+	} else if (k_sem_take(&probe_sem, HAWKBIT_SET_SERVER_TIMEOUT) == 0) {
+		snprintf(hb_cfg.server_port, sizeof(hb_cfg.server_port), "%u", port);
+		LOG_DBG("configured hawkbit server port: %s", hb_cfg.server_port);
```
this and others log messages: can potentially reuse the same `"hawkbit/server_port"` key used in the settings:

```c
LOG_DBG("configured %s: %s", "hawkbit/server_port", hb_cfg.server_port);
```

ycsin commented on subsys/mgmt/hawkbit/hawkbit.c:
```diff
@@ -826,8 +982,13 @@ static bool send_request(enum http_method method, enum hawkbit_http_request type
 	memset(&hb_context.recv_buf_tcp, 0, sizeof(hb_context.recv_buf_tcp));
 	hb_context.http_req.url = hb_context.url_buffer;
 	hb_context.http_req.method = method;
+	#ifdef CONFIG_HAWKBIT_SET_SETTINGS_RUNTIME
```
we dont really align `#ifdef` macros in Zephyr

ycsin reviewed (changes requested):
thanks for the patch, looks good in general, have you tested this? would be good to attach some logs.

also I think the samples should be updated accordingly

maass-hamburg committed:
samples: mgmt: hawkbit: runtime configuration for server address and port

Add runtime configuration for hawkbit server address and port

Signed-off-by: Fin Maaß <f.maass@vogl-electronic.com>

maass-hamburg commented:
> thanks for the patch, looks good in general, have you tested this? would be good to attach some logs.
> 
> also I think the samples should be updated accordingly
```
*** Booting Zephyr OS build v3.6.0-rc2-11-g10ac82c39ed8 ***
[00:00:00.011,000] <inf> main: Hawkbit sample app started
[00:00:00.011,000] <inf> main: Image build time: Mar  1 2024 14:26:45
[00:00:00.514,000] <inf> net_dhcpv4: Received: 10.101.66.10
[00:00:00.529,000] <inf> fs_nvs: 2 Sectors of 32768 bytes
[00:00:00.529,000] <inf> fs_nvs: alloc wra: 0, 7f88
[00:00:00.529,000] <inf> fs_nvs: data wra: 0, 8c
[00:00:00.529,000] <dbg> hawkbit: hawkbit_settings_set: <hawkbit/ddi_token> = 123456789
[00:00:00.529,000] <dbg> hawkbit: hawkbit_settings_set: <hawkbit/server_port> = 8070
[00:00:00.529,000] <dbg> hawkbit: hawkbit_settings_set: <hawkbit/server_addr> = 10.101.66.1
[00:00:00.529,000] <dbg> hawkbit: hawkbit_settings_set: <hawkbit/action_id> = 0
[00:00:00.529,000] <dbg> hawkbit: hawkbit_init: Action id: current 0
[00:00:00.529,000] <inf> hawkbit: Image is confirmed OK
[00:00:00.529,000] <dbg> hawkbit: hawkbit_set_server_addr: hawkbit/server_addr already correctly configured
[00:00:00.529,000] <dbg> hawkbit: hawkbit_set_server_port: hawkbit/server_port already correctly configured
[00:00:00.529,000] <inf> main: Starting hawkbit polling mode
[00:00:00.529,000] <dbg> hawkbit: hawkbit_settings_export: export hawkbit settings
[00:00:00.531,000] <inf> hawkbit: Polling target data from Hawkbit
[00:00:00.541,000] <dbg> hawkbit: hawkbit_probe: config.polling.sleep=00:05:00
[00:00:00.541,000] <inf> hawkbit: No update found
```

kartben commented on include/zephyr/mgmt/hawkbit.h:
```diff
@@ -63,6 +63,54 @@ void hawkbit_autohandler(void);
+ * @brief Set the Hawkbit server address.
+ *
+ * @param addr_str Server address to set.
+ * @return 0 on success.
+ * @return -EAGAIN if probe is currently running.
```
```suggestion
 * @retval 0 on success.
 * @retval -EAGAIN if probe is currently running.
```

kartben commented on include/zephyr/mgmt/hawkbit.h:
```diff
@@ -63,6 +63,54 @@ void hawkbit_autohandler(void);
+ * @brief Set the Hawkbit server port.
+ *
+ * @param port Server port to set.
+ * @return 0 on success.
+ * @return -EAGAIN if probe is currently running.
```
```suggestion
 * @retval 0 on success.
 * @retval -EAGAIN if probe is currently running.
```

kartben commented on include/zephyr/mgmt/hawkbit.h:
```diff
@@ -63,6 +63,54 @@ void hawkbit_autohandler(void);
+ * @brief Set the Hawkbit security token.
+ *
+ * @param token Security token to set.
+ * @return 0 on success.
+ * @return -EAGAIN if probe is currently running.
```
```suggestion
 * @retval 0 on success.
 * @retval -EAGAIN if probe is currently running.
```

kartben commented on samples/subsys/mgmt/hawkbit/overlay-runtine.conf:
there is a typo in the name of this file

kartben commented on subsys/mgmt/hawkbit/hawkbit.c:
```diff
@@ -799,6 +960,19 @@ static bool send_request(enum http_method method, enum hawkbit_http_request type
+#endif /* CONFIG_HAWKBIT_DDI_GATEWAY_SECURITY */
+
+	size_t s_header = strlen(header_auth) + DDI_SECURITY_TOKEN_SIZE + strlen("\r\n") + 1;
+	char *header = (char *)malloc(s_header);
```
don't think this is ever free()'d?

maass-hamburg replied:
it is now changed to a array, so we don’t need free()

ycsin commented on samples/subsys/mgmt/hawkbit/prj.conf:
```diff
@@ -14,6 +14,7 @@ CONFIG_HTTP_CLIENT=y
 CONFIG_DNS_RESOLVER=y
 CONFIG_JSON_LIBRARY=y
 CONFIG_BOOTLOADER_MCUBOOT=y
+CONFIG_SETTINGS=y
```
we probably don't need the `CONFIG_NVS=y` above anymore?

maass-hamburg replied:
We need a backend for the settings subsystem enabled. 
Therefore we use `CONFIG_NVS=y `. 
See https://docs.zephyrproject.org/latest/kconfig.html#CONFIG_SETTINGS_BACKEND

maass-hamburg commented:
ping @kartben

maass-hamburg commented:
ping @ycsin

ycsin commented:
> ping @ycsin

almost there..

maass-hamburg commented:
rebase, to resolve merge conflicts

ycsin commented on subsys/mgmt/hawkbit/hawkbit.c:
```diff
@@ -974,10 +1144,48 @@ static bool send_request(enum http_method method, enum hawkbit_http_request type
 }
 
+#ifdef CONFIG_HAWKBIT_SET_SETTINGS_RUNTIME
+static int check_hawkbit_server(void)
```
looks like this is configuring the default setting (Kconfig) instead?

maass-hamburg replied:
yes it is using them as a failback

ycsin commented on samples/subsys/mgmt/hawkbit/src/main.c:
```diff
@@ -36,6 +36,11 @@ int main(void)
 
 	ret = hawkbit_init();
 
+#ifdef CONFIG_HAWKBIT_SET_SETTINGS_RUNTIME
```
wouldn't this get set by `check_hawkbit_server`?

maass-hamburg replied:
I don't really understand what you mean

ycsin replied:
I mean isn't this change superfluous if it's setting it to the Kconfig value, since `check_hawkbit_server` is expected to use them as a fallback?

maass-hamburg replied:
yes it would be set by `check_hawkbit_server` and is redundant. But for a sample we want to show how to set the values, thats why its there.

ycsin replied:
fair enough

ycsin reviewed (commented):
The commit title & body of `mgmt: hawkbit: make server address and port settable during runtime` should be updated to include changes to the server auth token.

Another concern from me is that since we need different calls to change different config (addr, port, auth), what happens if the `autohandler` runs between the calls?

maass-hamburg commented:
> Another concern from me is that since we need different calls to change different config (addr, port, auth), what happens if the `autohandler` runs between the calls?

we have the `k_sem probe_sem` prohibiting the run. Changes can only be made if it is not taken.

ycsin commented:
> > Another concern from me is that since we need different calls to change different config (addr, port, auth), what happens if the `autohandler` runs between the calls?
> 
> we have the `k_sem probe_sem` prohibiting the run. Changes can only be made if it is not taken.

but it can't guarantee that the configuration (i.e. addr + auth) is atomic if the app/user has to change more than one config, what do you think about having a single function to configure all (server) settings with a single struct, i.e.

```c
struct hawkbit_server_config {
        char *addr;
        uint16_t port;
        char *auth_token;
};

int hawkbit_config_server(const struct hawkbit_server_config *const cfg); 
```

or another layer of lock to guarantee the atomicity of the configurations, otherwise it would mean that we expect the `autohandler` to fail if it happens to run between the call to configure the server address and the auth token.

maass-hamburg commented:
> > > Another concern from me is that since we need different calls to change different config (addr, port, auth), what happens if the `autohandler` runs between the calls?
> > 
> > 
> > we have the `k_sem probe_sem` prohibiting the run. Changes can only be made if it is not taken.
> 
> but it can't guarantee that the configuration (i.e. addr + auth) is atomic if the app/user has to change more than one config, what do you think about having a single function to configure all (server) settings with a single struct, i.e.
> 
> ```c
> struct hawkbit_server_config {
>         char *addr;
>         uint16_t port;
>         char *auth_token;
> };
> 
> int hawkbit_config_server(const struct hawkbit_server_config *const cfg); 
> ```
> 
> or another layer of lock to guarantee the atomicity of the configurations? otherwise it would mean that we expect the `autohandler` to fail if it happens to run between the call to configure the server address and the auth token.

I added `int hawkbit_config_server(char *server_addr, uint16_t server_port, char *auth_token);` for that, but without the use of a struct. It uses `k_sched_lock()` and `k_sched_unlock()` to prevent the `autohandler` to run in between.

ycsin commented:
> I added `int hawkbit_config_server(char *server_addr, uint16_t server_port, char *auth_token);` for that, but without the use of a struct. It uses `k_sched_lock()` and `k_sched_unlock()` to prevent the `autohandler` to run in between.

Are the other functions still necessary after having `hawkbit_config_server`? Otherwise the `hawkbit_config_server` can use the real `probe_sem` lock instead of locking the scheduler.

One last (probably) concern I have is that we are adding quite a few public functions which we have to take care of the deprecation process. If we use a config struct there might be less things to worry about.

maass-hamburg commented:
> hawkbit_config_server

I am willing to change the `hawkbit_config_server()` to use a struct.
But I would want to keep the functions public to just change one setting, as they provide a simple method to just change one value. They are also directly used by the shell and in the use case of the product I am part of developing, we also need the ability to change the server port and address independently of each other.

maass-hamburg commented:
@ycsin I changed it

ycsin commented:
> > > I added `int hawkbit_config_server(char *server_addr, uint16_t server_port, char *auth_token);` for that, but without the use of a struct. It uses `k_sched_lock()` and `k_sched_unlock()` to prevent the `autohandler` to run in between.
> > 
> > Are the other functions still necessary after having `hawkbit_config_server`? Otherwise the `hawkbit_config_server` can use the real `probe_sem` lock instead of locking the scheduler.
> > 
> > One last (probably) concern I have is that we are adding quite a few public functions which we have to take care of the deprecation process. If we use a config struct there might be less things to worry about.
> 
> I am willing to change the `hawkbit_config_server()` to use a struct.
> But I would want to keep the functions public to just change one setting, as they provide a simple method to just change one value. They are also directly used by the shell and in the use case of the product I am part of developing, we also need the ability to change the server port and address independently of each other.

Im on mobile now and haven't looked at the new code, I think the other way around is to have an internal implementation without locking probe sem. The public version will call these internal functions with the lock held. Then, the public functions that configure everything can lock and reuse these internal functions.

As for the individual configuration part, I think you can actually have the struct to configure a single setting, but setting other values to some impossible/invalid value so they wont be set by the function, i.e. 0 for port num and NULL for the addr or auth. But if you insist to have them split into multiple ones, I think that's acceptable as well..

ycsin commented on subsys/mgmt/hawkbit/hawkbit.c:
```diff
@@ -598,6 +683,98 @@ static void hawkbit_dump_deployment(struct hawkbit_dep_res *d)
 	LOG_DBG("%s=%s", "md5sum-http", l->md5sum_http.href);
 }
 
+#ifdef CONFIG_HAWKBIT_SET_SETTINGS_RUNTIME
```
This looks incredibly complex, I was expecting something simpler, will something like the following achieve what you want?

```suggestion
#ifdef CONFIG_HAWKBIT_SET_SETTINGS_RUNTIME
static void hawkbit_config_server_addr_nolock(const char *const addr)
{
	if (!strcmp(addr, hb_cfg.server_addr)) {
		LOG_DBG("%s already correctly configured", "hawkbit/server_addr");
		return;
	}

	strncpy(hb_cfg.server_addr, addr, sizeof(hb_cfg.server_addr));
	LOG_DBG("configured %s: %s", "hawkbit/server_addr", hb_cfg.server_addr);
}

static void hawkbit_config_server_port_nolock(const uint16_t port)
{
	if (port == atoi(hb_cfg.server_port)) {
		LOG_DBG("%s already correctly configured", "hawkbit/server_port");
		return;
	}

	snprintf(hb_cfg.server_port, sizeof(hb_cfg.server_port), "%u", port);
	LOG_DBG("configured %s: %s", "hawkbit/server_port", hb_cfg.server_port);
}

#ifndef CONFIG_HAWKBIT_DDI_NO_SECURITY
static void hawkbit_config_server_auth_nolock(const char *const auth)
{
	if (!strcmp(auth, hb_cfg.ddi_security_token)) {
		LOG_DBG("%s already correctly configured", "hawkbit/ddi_token");
	}

	strncpy(hb_cfg.ddi_security_token, auth, sizeof(hb_cfg.ddi_security_token));
	LOG_DBG("configured %s: %s", "hawkbit/ddi_token", hb_cfg.ddi_security_token);
}
#endif /* CONFIG_HAWKBIT_DDI_NO_SECURITY */

int hawkbit_config_server(struct hawkbit_set_config *set_config)
{
	if (k_sem_take(&probe_sem, HAWKBIT_SET_SERVER_TIMEOUT) != 0) {
		LOG_DBG("Failed to configure server");
		return -EAGAIN;
	}

	if (set_config->addr != NULL) {
		hawkbit_config_server_addr_nolock(set_config->addr);
	}

	if (set_config->port != 0U) {
		hawkbit_config_server_port_nolock(set_config->port);
	}

#ifndef CONFIG_HAWKBIT_DDI_NO_SECURITY
	if (set_config->auth != NULL) {
		hawkbit_config_server_auth_nolock(set_config->auth);
	}
#endif /* CONFIG_HAWKBIT_DDI_NO_SECURITY */

	settings_save();

	k_sem_give(&probe_sem);

	return 0;
}
#endif /* CONFIG_HAWKBIT_SET_SETTINGS_RUNTIME */
```

Public functions like `hawkbit_set_server_addr` can call `hawkbit_config_server_addr_nolock` with the lock held, arg check can go into the internal functions if desired, they can be made to return error codes if necessary.

maass-hamburg replied:
Well I wanted to do the check if a value is already correctly configured before taking the lock, so that if nothing changes taking the lock is not required.

ycsin replied:
I'd expect that the function is called only when there's something that needs to be changed, if an application is calling the config function repeatedly without the need to change anything then the application has to be fixed.

maass-hamburg replied:
I will remove the part where it will compare the old and new value

maass-hamburg replied:
@ycsin it's changed now

ycsin reviewed (commented):
If we have something like `hawkbit_set_server_config`, I think we can have `hawkbit_get_server_config` to match?

maass-hamburg commented:
> If we have something like `hawkbit_set_server_config`, I think we can have `hawkbit_get_server_config` to match?

Sure

maass-hamburg committed:
doc: releases: 3.7: change of hawkbit

Mention change of hawkbit in the migration guide and the release notes.

Signed-off-by: Fin Maaß <f.maass@vogl-electronic.com>

ycsin commented on doc/releases/migration-guide-3.7.rst:
```diff
@@ -230,6 +230,14 @@ Networking
+hawkBit
+=======
+
+  * :kconfig:option:`CONFIG_HAWKBIT_PORT` is now a int instead of a string.
```
```suggestion
  * :kconfig:option:`CONFIG_HAWKBIT_PORT` is now an int instead of a string.
```

ycsin commented on subsys/mgmt/hawkbit/hawkbit.c:
```diff
@@ -194,6 +190,40 @@ static const struct json_obj_descr json_dep_fbk_descr[] = {
+		return rc;
+	}
+
+
```
```suggestion
```

ycsin commented on subsys/mgmt/hawkbit/hawkbit.c:
```diff
@@ -556,34 +587,12 @@ static void hawkbit_dump_deployment(struct hawkbit_dep_res *d)
-		LOG_ERR("Failed to mount storage flash: %d", rc);
-		return rc;
-	}
+	settings_subsys_init();
+	settings_load_subtree("hawkbit");
```
what would happen if setting fails to init/load?

maass-hamburg replied:
now handeled

ycsin commented on subsys/mgmt/hawkbit/hawkbit.c:
```diff
@@ -585,6 +694,50 @@ static void hawkbit_dump_deployment(struct hawkbit_dep_res *d)
+}
+#endif /* CONFIG_HAWKBIT_SET_SETTINGS_RUNTIME */
+
+struct hawkbit_runtime_config hawkbit_get_config(void)
+{
+	struct hawkbit_runtime_config config = {
+		.server_addr = HAWKBIT_SERVER,
+		.server_port = HAWKBIT_PORT_INT,
+		.auth_token = HAWKBIT_DDI_SECURITY_TOKEN,
+	};
+
+	return config;
+}
```
I reckon it's more efficient to let the caller to pass the struct into this function by reference, instead of making copies

ycsin replied:
Yeah, hopefully it doesnt grow too much

ycsin commented on subsys/mgmt/hawkbit/shell.c:
```diff
@@ -78,14 +79,86 @@ static int cmd_info(const struct shell *sh, size_t argc, char *argv)
+
+	shell_fprintf(sh, SHELL_INFO, "Init hawkBit ...\n");
+
+	hawkbit_init();
```
risky if user calls this more than once

maass-hamburg replied:
`hawkbit_init()` now includes a check if it has already been initialised

ycsin commented on subsys/mgmt/hawkbit/hawkbit.c:
```diff
@@ -556,53 +589,51 @@ static void hawkbit_dump_deployment(struct hawkbit_dep_res *d)
-	if (!device_is_ready(fs.flash_device)) {
-		LOG_ERR("Flash device not ready");
-		return -ENODEV;
+	if (hawkbit_initialized) {
```
This should go into another commit

ycsin commented on subsys/mgmt/hawkbit/hawkbit.c:
```diff
@@ -556,53 +589,51 @@ static void hawkbit_dump_deployment(struct hawkbit_dep_res *d)
-	if (rc) {
-		LOG_ERR("Failed to get storage page info: %d", rc);
-		return -EIO;
+	ret = k_sem_take(&probe_sem, K_NO_WAIT);
```
Why is this required?

ycsin commented on subsys/mgmt/hawkbit/hawkbit.c:
```diff
@@ -969,6 +1000,10 @@ enum hawkbit_response hawkbit_probe(void)
 	     deployment_base[DEPLOYMENT_BASE_SIZE] = { 0 },
 	     firmware_version[BOOT_IMG_VER_STRLEN_MAX] = { 0 };
 
+	if (!hawkbit_initialized) {
```
should be in another commit

ycsin commented on subsys/mgmt/hawkbit/hawkbit.c:
```diff
@@ -1227,6 +1260,10 @@ static void autohandler(struct k_work *work)
 		LOG_INF("Metadata error");
 		break;
 
+	case HAWKBIT_NOT_INITIALIZED:
```
Should be in another commit

ycsin reviewed (changes requested):
should update testcase?

maass-hamburg committed:
mgmt: hawkbit: migrate to settings api

Migrates from nvs to settings api.

Signed-off-by: Fin Maaß <f.maass@vogl-electronic.com>

maass-hamburg committed:
mgmt: hawkbit: check for init

check for init before running hawkbit.

Signed-off-by: Fin Maaß <f.maass@vogl-electronic.com>

maass-hamburg committed:
samples: mgmt: hawkbit: add testcase

Add sample testcase for settings. that are
set durring runtime.

Signed-off-by: Fin Maaß <f.maass@vogl-electronic.com>

maass-hamburg commented:
rebased to resolve conflicts

ycsin commented on subsys/mgmt/hawkbit/shell.c:
```diff
@@ -4,6 +4,7 @@
  * SPDX-License-Identifier: Apache-2.0
  */
 
+#include <stdlib.h>
```
is this required?

maass-hamburg replied:
yes for `atoi()` in line 129

ycsin reviewed (approved):
Good work, LGTM, thanks!

real-tintin commented on subsys/mgmt/hawkbit/hawkbit.c:
```diff
@@ -577,37 +720,72 @@ int hawkbit_default_config_data_cb(const char *device_id, uint8_t *buffer, const
+#ifdef CONFIG_HAWKBIT_SET_SETTINGS_RUNTIME
+int hawkbit_set_config(struct hawkbit_runtime_config *config)
+{
+	if (k_sem_take(&probe_sem, HAWKBIT_SET_SERVER_TIMEOUT) == 0) {
```
Consider K_NO_WAIT instead. Imo it makes sense to call hawkbit_set_config prior to any probe/autoprobe.

maass-hamburg replied:
I have no objection against it. Due to it already being merged, please open a PR to have it changed.

maass-hamburg commented:
> @maass-hamburg could you rebase this on the earlier changes? Then I can test it. We use the dynamic setting.

@real-tintin this had been merged months ago, I think you wrote on the wrong PR.
